File: galite-core/src/main/java/org/kopi/vkopi/lib/ui/swing/form/DChartBlock.java
/*
 * Copyright (c) 1990-2021 kopiRight Managed Solutions GmbH
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License version 2.1 as published by the Free Software Foundation.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 *
 * $Id: DChartBlock.java 34997 2016-12-01 09:51:43Z hacheni $
 */

package org.kopi.vkopi.lib.ui.swing.form;

import java.awt.AWTEvent;
import java.awt.Adjustable;
import java.awt.Toolkit;
import java.awt.event.AdjustmentEvent;
import java.awt.event.AdjustmentListener;
import java.awt.event.InvocationEvent;

import javax.swing.JScrollBar;

import org.kopi.galite.visual.form.LayoutManager;
import org.kopi.galite.visual.form.VBlock;
import org.kopi.vkopi.lib.ui.swing.visual.SwingThreadHandler;
import org.kopi.galite.visual.visual.Action;
import org.kopi.galite.visual.visual.VException;

public class DChartBlock extends DBlock {

  /*
   * ----------------------------------------------------------------------
   * CONSTRUCTION
   * ----------------------------------------------------------------------
   */

  public DChartBlock(DForm form, VBlock model) {
    super(form, model);
    if (getModel().getDisplaySize() < getModel().getBufferSize()) {
      scrollBar = createScrollBar();
      scrollBar.setFocusable(false);
      addScrollBar(scrollBar);
    }
  }

  protected void addScrollBar(JScrollBar bar) {
    add(bar);
  }

  protected JScrollBar createScrollBar() {
    final JScrollBar          scrollBar;

    scrollBar = new JScrollBar(Adjustable.VERTICAL,
                               0,
                               getModel().getDisplaySize(),
                               0,
                               getModel().getBufferSize() - 1);

    adjustmentListener = new AdjustmentListener() {
        public void adjustmentValueChanged(final AdjustmentEvent e) {
          if (((DForm)getFormView()).getInAction()) {
            // do not change the rows if there is currently a
            // another command executed
            return;
          }
          getFormView().performAsyncAction(new Action("chart") {
              public void execute() throws VException {
                if (!init) {
                  init = true;
                } else {
                  // Must be outside of event disp thread, so that
                  // the scrollbar loses focuses and stop the scroll thread
                  try {
                    setScrollPos(e.getValue());
                  } catch (VException e) {
                    Toolkit.getDefaultToolkit().getSystemEventQueue().postEvent(new RNCEvent());
                    throw e;
                  }
                }
              }
            }); // performBasicAction : not asyncron!
        }
        private boolean		init;
      };
    return scrollBar;
  }

  protected LayoutManager createLayoutManager() {
    return new KopiMultiBlockLayout(displayedFields, getModel().getDisplaySize() + 1 /* Records + Header*/);
  }


  public void validRecordNumberChanged() {
    if (getModel().getDisplaySize() < getModel().getBufferSize()) {
      // only useful if there is a scrollbar

      Toolkit.getDefaultToolkit().getSystemEventQueue().postEvent(new RNCEvent());
    }
  }

  protected void refresh(boolean force) {
    SwingThreadHandler.verifyRunsInEventThread("DBlock refresh");

    super.refresh(force);
    if (scrollBar != null) {
      updateScrollbar();
    }
  }

  private void updateScrollbar() {
    SwingThreadHandler.verifyRunsInEventThread("DChartBlock updateScrollbar");
    int         validRecords = getModel().getNumberOfValidRecord();
    int         dispSize     = getModel().getDisplaySize();

    if (validRecords > dispSize) {
      scrollBar.removeAdjustmentListener(adjustmentListener);
      scrollBar.setValues(getModel().getNumberOfValidRecordBefore(getRecordFromDisplayLine(0)),
                          dispSize,
                          0,
                          validRecords);
      scrollBar.addAdjustmentListener(adjustmentListener);
      scrollBar.setEnabled(true);
    } else {
      scrollBar.setEnabled(false);
    }
  }

  // ----------------------------------------------------------------------
  // Event Handling
  // ----------------------------------------------------------------------

  class RNCEvent extends InvocationEvent implements Runnable {

	RNCEvent() {
      super(DChartBlock.this, EVT_RNC, null, null, false);
      runnable = this;
    }

    public void run() {
      updateScrollbar();
    }

    static final int    EVT_RNC = 2100;
    /**
	 * Comment for <code>serialVersionUID</code>
	 */
	private static final long serialVersionUID = 1727735012273500143L;
  }

  protected AWTEvent coalesceEvents(AWTEvent existingEvent,
                                    AWTEvent newEvent) {
    int                 id = existingEvent.getID();

    switch (id) {
    case RNCEvent.EVT_RNC:
      // coalesce updates to the scrollbar
      return existingEvent;
    default:
      return super.coalesceEvents(existingEvent, newEvent);
    }
  }

  // ----------------------------------------------------------------------
  // PRIVATE DATA
  // ----------------------------------------------------------------------

  private AdjustmentListener    adjustmentListener;
  private JScrollBar            scrollBar;
  /**
	 * Comment for <code>serialVersionUID</code>
	 */
  private static final long serialVersionUID = -7342909071556158313L;
}

File: galite-core/src/main/kotlin/org/kopi/galite/ui/vaadin/form/DChartBlock.kt
/*
 * Copyright (c) 2013-2020 kopiLeft Services SARL, Tunis TN
 * Copyright (c) 1990-2020 kopiRight Managed Solutions GmbH, Wien AT
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License version 2.1 as published by the Free Software Foundation.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */
package org.kopi.galite.ui.vaadin.form

import org.kopi.galite.base.UComponent
import org.kopi.galite.form.Alignment
import org.kopi.galite.form.VBlock
import org.kopi.galite.form.VConstants
import org.kopi.galite.ui.vaadin.block.BlockLayout
import org.kopi.galite.ui.vaadin.block.BlockListener
import org.kopi.galite.ui.vaadin.block.ChartBlockLayout
import org.kopi.galite.visual.Action
import org.kopi.galite.visual.MessageCode
import org.kopi.galite.visual.VException
import org.kopi.galite.visual.VExecFailedException

import com.vaadin.flow.component.Component

/**
 * The `DChartBlock` is a [DBlock] representing
 * a chart block where data is represented in a data grid.
 *
 * @param parent The parent form.
 * @param model The block model.
 */
open class DChartBlock(parent: DForm, model: VBlock) : DBlock(parent, model), BlockListener {

  //-------------------------------------------------
  // DATA MEMBERS
  //-------------------------------------------------
  private var init = false

  /*
   * This flag was added to avoid mutual communication
   * between client and server side when the scroll bar
   * position is changed from the client side.
   * Thus, scroll bar position is not changed by server
   * side when it is changed from the client side
   * @see {@link #refresh(boolean)
   */
  private var scrolling = false

  init {
    if (model.displaySize < model.bufferSize) {
      addBlockListener(this)
    }
  }

  //---------------------------------------------------
  // IMPLEMENTATIONS
  //---------------------------------------------------
  override fun add(comp: UComponent?, constraints: Alignment) {
    addComponent(comp as? Component,
                 constraints.x,
                 constraints.y,
                 constraints.width,
                 constraints.height,
                 constraints.alignRight,
                 constraints.useAll)
  }

  override fun createLayout(): BlockLayout {
    val layout = ChartBlockLayout(displayedFields, model.displaySize + 1)

    layout.hasScroll = model.displaySize < model.bufferSize
    return layout
  }

  override fun validRecordNumberChanged() {
    if (model.displaySize < model.bufferSize) {
      updateScrollbar()
    }
  }

  override fun recordInfoChanged(rec: Int, info: Int) {
    // send the new records info
    fireRecordInfoChanged(rec, info)
    model.updateColor(rec)
  }

  override fun refresh(force: Boolean) {
    super.refresh(force)
    // don't update scroll bar when the scroll position
    // is being changing from the client side.
    if (!scrolling) {
      updateScrollbar()
    }
  }

  override fun onScroll(value: Int) {
    if ((getFormView() as DForm).inAction) {
      // do not change the rows if there is currently a
      // another command executed
      return
    }
    if (!init) {
      init = true // on initialization, we do not scroll.
    } else {
      try {
        scrolling = true
        setScrollPos(value)
        scrolling = false
      } catch (e: VException) {
        e.printStackTrace()
      }
    }
  }

  override fun onActiveRecordChange(record: Int, sortedTopRec: Int) {
    if (record != model.activeRecord) {
      model.form.performAsyncAction(object : Action() {
        override fun execute() {
          // go to the correct block if necessary
          if (model !== model.form.getActiveBlock()) {
            if (!model.isAccessible) {
              throw VExecFailedException(MessageCode.getMessage("VIS-00025"))
            }
            model.form.gotoBlock(model)
          }
          if (fireTriggers()) {
            sortedToprec = sortedTopRec
            model.gotoRecord(record)
          } else {
            model.activeRecord = record
            model.currentRecord = record
            sortedToprec = sortedTopRec
            refresh(false)
          }
        }

        /**
         * Returns `true` if there some triggers that should be fired
         * when changing the active record of the block model.
         * @return `true` if there some triggers that should be fired.
         * @throws VException Visual errors.
         */
        private fun fireTriggers(): Boolean {
          return isActiveBlock && (fireTriggerOnBlock() || fireTriggersOnActiveField())
        }

        /**
         * Returns `true` if this block is the active block in the form.
         * @return `true` if this block is the active block in the form.
         * @throws VException Visual errors.
         */
        private val isActiveBlock: Boolean
          get() = (model.form.getActiveBlock() != null
                  && model.form.getActiveBlock() == model)

        /**
         * Returns `true` is there are trigger that should be fired on this block.
         * @return `true` is there are trigger that should be fired on this block.
         * @throws VException Visual errors.
         */
        private fun fireTriggerOnBlock(): Boolean {
          return (model.hasTrigger(VConstants.TRG_PREREC)
                  || model.hasTrigger(VConstants.TRG_VALREC))
        }

        /**
         * Returns `true` is there are trigger that should be fired on the active field.
         * @return `true` is there are trigger that should be fired on the active field.
         * @throws VException Visual errors.
         */
        private fun fireTriggersOnActiveField(): Boolean {
          return if (model.activeField != null) {
            (model.activeField!!.hasTrigger(VConstants.TRG_POSTCHG)
                    || model.activeField!!.hasTrigger(VConstants.TRG_VALFLD)
                    || model.activeField!!.hasTrigger(VConstants.TRG_FORMAT)
                    || model.activeField!!.hasTrigger(VConstants.TRG_PREVAL)
                    || model.activeField!!.hasTrigger(VConstants.TRG_POSTFLD))
          } else {
            false
          }
        }
      })
    }
  }

  /**
   * Updates the scroll bar position.
   */
  override fun updateScrollbar() {
    //BackgroundThreadHandler.access(Runnable { TODO
      val validRecords = model.numberOfValidRecord
      val dispSize = model.displaySize

      updateScroll(dispSize,
                   validRecords,
                   validRecords > dispSize,
                   model.getNumberOfValidRecordBefore(getRecordFromDisplayLine(0)))
    //})
  }
}
